### PR TITLE
[StorageMover] Add new module Az.StorageMover

### PR DESCRIPTION
**Pull request by yifanz0**
## Description

* First preview release for module Az.StorageMover. Cmdlets added include:
    - `Get-AzStorageMover`
    - `Get-AzStorageMoverAgent`
    - `Get-AzStorageMoverEndpoint`
    - `Get-AzStorageMoverJobDefinition`
    - `Get-AzStorageMoverJobRun`
    - `Get-AzStorageMoverProject`
    - `New-AzStorageMover`
    - `New-AzStorageMoverAzStorageContainerEndpoint`
    - `New-AzStorageMoverJobDefinition`
    - `New-AzStorageMoverNfsEndpoint`
    - `New-AzStorageMoverProject`
    - `Remove-AzStorageMover`
    - `Remove-AzStorageMoverEndpoint`
    - `Remove-AzStorageMoverJobDefinition`
    - `Remove-AzStorageMoverProject`
    - `Start-AzStorageMoverJobDefinition`
    - `Stop-AzStorageMoverJobDefinition`
    - `Unregister-AzStorageMoverAgent`
    - `Update-AzStorageMover`
    - `Update-AzStorageMoverAgent`
    - `Update-AzStorageMoverAzStorageContainerEndpoint`
    - `Update-AzStorageMoverJobDefinition`
    - `Update-AzStorageMoverNfsEndpoint`
    - `Update-AzStorageMoverProject`

Design link: https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/1234

## Checklist

- [x] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** select appropriate branch. Cmdlets from Autorest.PowerShell should go to [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch. 
* **SHOULD** make the title of PR clear and informative, and in the present imperative tense. 
* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. Add changelog in description section if PR goes into [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** have approved design review for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) ([_Microsoft internal only_](../blob/main/CONTRIBUTING.md#onboarding)) with following situations
    * Create new module from the scratch
    * Create new resource types which are not easy to conform to [Azure PowerShell Design Guidelines](../blob/main/documentation/development-docs/design-guidelines)
    * Create new resource type which name doesn't use module name as prefix
    * Have design question before implementation
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** introduce [breaking changes](../blob/main/documentation/breaking-changes/breaking-changes-definition.md) in Az minor release except preview version.
* **SHOULD NOT** change version of module in pull request

### CODE REVIEW

File: src/StorageMover/README.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+require:
+# readme.azure.noprofile.md is the common configuration file
+  - $(this-folder)/../readme.azure.noprofile.md
+input-file:
+# You need to specify your swagger files here.
+#  - https://github.com/Azure/azure-rest-api-specs-pr/blob/RPSaaSMaster/specification/storagemover/resource-manager/Microsoft.StorageMover/preview/2022-07-01-preview/storagemover.json
+
+# If the swagger has not been put in the repo, you may uncomment the following line and refer to it locally
+# - D:\code\swaggerrepo\specification\storagemover\resource-manager\Microsoft.StorageMover\preview\2022-07-01-preview\storagemover.json
+ - $(this-folder)/../StorageMoverSwagger/swagger/specification/storagemover/resource-manager/Microsoft.StorageMover/preview/2022-07-01-preview/storagemover.json
```
Comment (BethanyZhou) [inline src/StorageMover/README.md:41]:
```suggestion
require:
# readme.azure.noprofile.md is the common configuration file
  - $(this-folder)/../readme.azure.noprofile.md
input-file:
# You need to specify your swagger files here.
#  - https://github.com/Azure/azure-rest-api-specs-pr/blob/RPSaaSMaster/specification/storagemover/resource-manager/Microsoft.StorageMover/preview/2022-07-01-preview/storagemover.json

# If the swagger has not been put in the repo, you may uncomment the following line and refer to it locally
# - D:\code\swaggerrepo\specification\storagemover\resource-manager\Microsoft.StorageMover\preview\2022-07-01-preview\storagemover.json
  - $(this-folder)/../StorageMoverSwagger/swagger/specification/storagemover/resource-manager/Microsoft.StorageMover/preview/2022-07-01-preview/storagemover.json
```

File: src/StorageMover/README.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+## Authentication
+AutoRest does not generate authentication code for the module. Authentication is handled via Az.Accounts by altering the HTTP payload before it is sent.
+
+## Development
+For information on how to develop for `Az.StorageMover`, see [how-to.md](how-to.md).
+<!-- endregion -->
+
+``` yaml
+# Please specify the commit id that includes your features to make sure generated codes stable.
+branch: main
```
Comment (BethanyZhou) [inline src/StorageMover/README.md:31]:
```suggestion
```
Need to remove as we are using local file

File: src/StorageMover/README.md (hunk 1 above)
Comment (BethanyZhou) [inline src/StorageMover/README.md:41]:
```suggestion
  - $(this-folder)/../StorageMoverSwagger/swagger/specification/storagemover/resource-manager/Microsoft.StorageMover/preview/2022-07-01-preview/storagemover.json
```

**Comment by BethanyZhou** [comment]
Convert this PR to draft PR as current is using local swagger to test CI. If the swagger of storage mover changes to public available, please remove local file and mark this PR to ready to review.

**Review by BethanyZhou — CHANGES_REQUESTED** [review]

File: src/StorageMover/custom/NewAzStorageMoverAzStorageContainerEndpoint.ps1 [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        # Use the default credentials for the proxy
+        ${ProxyUseDefaultCredentials}
+    )
+
+    process {
+        $Properties = [Microsoft.Azure.PowerShell.Cmdlets.StorageMover.Models.Api20220701Preview.AzureStorageBlobContainerEndpointProperties]::New()
+
+        if ($PSBoundParameters.ContainsKey('BlobContainerName')) {
+            $Properties.BlobContainerName = $BlobContainerName
+        }
```
[the remark this answers is not in the document]
Comment (blueww) [reply src/StorageMover/custom/NewAzStorageMoverAzStorageContainerEndpoint.ps1:138]: @BethanyZhou 
I believe we should remove the parameter from `$PSBoundParameters`, for all parameters won't input to `Az.StorageMover.internal\New-AzStorageMoverEndpoint`. 

Besides that, for "combine", what do you mean? Do you mean we should combine different parameters to one if statement, or add the remove parameter from `$PSBoundParameters` for each property parameter?

[the remark this answers is not in the document]
Comment (blueww) [reply src/StorageMover/custom/NewAzStorageMoverAzStorageContainerEndpoint.ps1:138]: Never mind, I get what it means.

File: src/StorageMover/custom/NewAzStorageMoverAzStorageContainerEndpoint.ps1 [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+        [Parameter(DontShow)]
+        [Microsoft.Azure.PowerShell.Cmdlets.StorageMover.Category('Runtime')]
+        [System.Management.Automation.SwitchParameter]
+        # Use the default credentials for the proxy
+        ${ProxyUseDefaultCredentials}
+    )
+
+    process {
+        $Properties = [Microsoft.Azure.PowerShell.Cmdlets.StorageMover.Models.Api20220701Preview.AzureStorageBlobContainerEndpointProperties]::New()
```
[the remark this answers is not in the document]
Comment (yifanz0) [reply src/StorageMover/custom/NewAzStorageMoverAzStorageContainerEndpoint.ps1:134]: The EndpointType is set a bit later on line 146

**Comment by blueww** [comment]
@BethanyZhou 

The static analysis fail, following is error , see [link ](https://dev.azure.com/azure-sdk/public/_build/results?buildId=1727087&view=logs&jobId=e1b7b984-9f58-529f-7c5a-f15f8e35cfa6&j=8aaed9ad-5f17-55b5-7c72-2ac5707cbf7b&t=96d7b2a4-f64e-595d-e325-a9fcd671d454).
I have run it 3 times , always fail.
The error looks not related with our code change. 
Do you have any idea how to fix it?
```
C:\hostedtoolcache\windows\dotnet\sdk\3.1.421\Microsoft.Common.CurrentVersion.targets(4364,5): warning MSB3026: Could not copy "C:\Users\VssAdministrator\.nuget\packages\microsoft.azure.powershell.clients.websites\1.3.59-preview\lib\netstandard2.0\Microsoft.Azure.PowerShell.Clients.Websites.dll" to "D:\a\1\s\artifacts/Debug\Az.Accounts\Microsoft.Azure.PowerShell.Clients.Websites.dll". Beginning retry 1 in 1000ms. The process cannot access the file 'D:\a\1\s\artifacts\Debug\Az.Accounts\Microsoft.Azure.PowerShell.Clients.Websites.dll' because it is being used by another process.  [D:\a\1\s\src\Accounts\AuthenticationAssemblyLoadContext\AuthenticationAssemblyLoadContext.csproj] [D:\a\1\s\build.proj]
C:\hostedtoolcache\windows\dotnet\sdk\3.1.421\Microsoft.Common.CurrentVersion.targets(4364,5): warning MSB3026: Could not copy "C:\Users\VssAdministrator\.nuget\packages\microsoft.azure.powershell.clients.resourcemanager\1.3.59-preview\lib\netstandard2.0\Microsoft.Azure.PowerShell.Clients.ResourceManager.dll" to "D:\a\1\s\artifacts/Debug\Az.Accounts\Microsoft.Azure.PowerShell.Clients.ResourceManager.dll". Beginning retry 1 in 1000ms. The process cannot access the file 'D:\a\1\s\artifacts\Debug\Az.Accounts\Microsoft.Azure.PowerShell.Clients.ResourceManager.dll' because it is being used by another process.  [D:\a\1\s\src\Accounts\AuthenticationAssemblyLoadContext\AuthenticationAssemblyLoadContext.csproj] [D:\a\1\s\build.proj]
C:\hostedtoolcache\windows\dotnet\sdk\3.1.421\Microsoft.Common.CurrentVersion.targets(4364,5): warning MSB3026: Could not copy "C:\Users\VssAdministrator\.nuget\packages\microsoft.azure.powershell.clients.storage.management\1.3.59-preview\lib\netstandard2.0\Microsoft.Azure.PowerShell.Clients.Storage.Management.dll" to "D:\a\1\s\artifacts/Debug\Az.Accounts\Microsoft.Azure.PowerShell.Clients.Storage.Management.dll". Beginning retry 1 in 1000ms. The process cannot access the file 'D:\a\1\s\artifacts\Debug\Az.Accounts\Microsoft.Azure.PowerShell.Clients.Storage.Management.dll' because it is being used by another process.  [D:\a\1\s\src\Accounts\AuthenticationAssemblyLoadContext\AuthenticationAssemblyLoadContext.csproj] [D:\a\1\s\build.proj]
C:\hostedtoolcache\windows\dotnet\sdk\3.1.421\Microsoft.Common.CurrentVersion.targets(2084,5): warning MSB3245: Could not resolve this reference. Could not locate the assembly "FuzzySharp". Check to make sure the assembly exists on disk. If this reference is required by your code, you may get compilation errors. [D:\a\1\s\src\Accounts\Accounts\Accounts.csproj] [D:\a\1\s\build.proj]
Utilities\CommandNotFoundHelper.cs(24,7): error CS0246: The type or namespace name 'FuzzySharp' could not be found (are you missing a using directive or an assembly reference?) [D:\a\1\s\src\Accounts\Accounts\Accounts.csproj] [D:\a\1\s\build.proj]
Utilities\CommandNotFoundHelper.cs(25,7): error CS0246: The type or namespace name 'FuzzySharp' could not be found (are you missing a using directive or an assembly reference?) [D:\a\1\s\src\Accounts\Accounts\Accounts.csproj] [D:\a\1\s\build.proj]

```

File: src/StorageMover/README.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+# For new RP, the version is 0.1.0
+module-version: 0.1.0
+# Normally, title is the service name
+title: StorageMover
+subject-prefix: $(service-name)
+nested-object-to-string: true
+identity-correction-for-post: true 
+# If there are post APIs for some kinds of actions in the RP, you may need to 
+# uncomment following line to support viaIdentity for these post APIs
+# identity-correction-for-post: true
```
[the remark this answers is not in the document]
Comment (yifanz0) [reply src/StorageMover/README.md:50]: Done.

File: src/StorageMover/docs/Get-AzStorageMoverEndpoint.md [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+SystemDataCreatedByType      : User
+SystemDataLastModifiedAt     : 7/18/2022 7:28:30 AM
+SystemDataLastModifiedBy     : xxxxxxx
+SystemDataLastModifiedByType : User
+Type                         : microsoft.storagemover/storagemovers/endpoints
+```
+
+This command gets all the endpoints under a Storage mover.
+
+### Example 2: {{ Add title here }}
```
[the remark this answers is not in the document]
Comment (yifanz0) [reply src/StorageMover/docs/Get-AzStorageMoverEndpoint.md]: Done

File: src/StorageMover/custom/UnregisterAzStorageMoverAgent.ps1 [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+  [ProjectName<String>]: The name of the project resource.
+  [ResourceGroupName<String>]: The name of the resource group. The name is case insensitive.
+  [StorageMoverName<String>]: The name of the Storage Mover resource.
+  [SubscriptionId<String>]: The ID of the target subscription.
+.Link
+https://docs.microsoft.com/powershell/module/az.storagemover/remove-azstoragemoveragent
+#>
+function Unregister-AzStorageMoverAgent {
+    [OutputType([System.Boolean])]
+[CmdletBinding(DefaultParameterSetName = 'Delete', PositionalBinding =$false, SupportsShouldProcess, ConfirmImpact = 'Medium')]
```
[the remark this answers is not in the document]
Comment (yifanz0) [reply src/StorageMover/custom/UnregisterAzStorageMoverAgent.ps1]: Done

File: src/StorageMover/custom/UpdateAzStorageMoverAzStorageContainerEndpoint.ps1 [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+  [ResourceGroupName<String>]: The name of the resource group. The name is case insensitive.
+  [StorageMoverName<String>]: The name of the Storage Mover resource.
+  [SubscriptionId<String>]: The ID of the target subscription.
+
+.Link
+https://docs.microsoft.com/powershell/module/az.storagemover/update-azstoragemoverendpoint
+#>
+function Update-AzStorageMoverAzStorageContainerEndpoint {
+    [OutputType([Microsoft.Azure.PowerShell.Cmdlets.StorageMover.Models.Api20220701Preview.IEndpoint])]
+[CmdletBinding(DefaultParameterSetName = 'UpdateExpanded', PositionalBinding =$false, SupportsShouldProcess, ConfirmImpact = 'Medium')]
```
[the remark this answers is not in the document]
Comment (yifanz0) [reply src/StorageMover/custom/UpdateAzStorageMoverAzStorageContainerEndpoint.ps1]: Done

File: src/StorageMover/examples/Update-AzStorageMoverProject.md [hunk 9]
```diff
@@ -0,0 +1,21 @@
+### Example 1: Update a project
+```powershell
+Update-AzStorageMoverProject -ResourceGroupName myResourceGroup -StorageMoverName myStorageMover -Name myProject -Description "Update Description""
```
Comment (isra-fel) [inline src/StorageMover/examples/Update-AzStorageMoverProject.md:3]:
```suggestion
Update-AzStorageMoverProject -ResourceGroupName myResourceGroup -StorageMoverName myStorageMover -Name myProject -Description "Update Description"
```

File: src/StorageMover/examples/New-AzStorageMoverJobDefinition.md [hunk 10]
```diff
@@ -0,0 +1,34 @@
+### Example 1: Create a job definition
+```powershell
+New-AzStorageMoverJobDefinition -Name myJob -ProjectName myProject -ResourceGroupName myResourceGroup -StorageMoverName myStorageMover -AgentName myAgent -SourceName myNfsEndpoint -TargetName myContainerEndpoint -CopyMode "Additive" -Description "job definition"
+``
```
Comment (isra-fel) [inline src/StorageMover/examples/New-AzStorageMoverJobDefinition.md:4]:
````suggestion
```
````

File: src/StorageMover/README.md [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+      verb: Set
+    remove: true
+  - where:
+      model-name: StorageMover
+    set:
+      suppress-format: true
+  - where:
+      model-name: JobDefinition
+    set:
+      suppress-format: true
```
Comment (isra-fel) [inline src/StorageMover/README.md:60]: `model-name` is actually regular expression, so you can group the same directives together like:
```suggestion
  - where:
      model-name: StorageMover|JobDefinition
    set:
      suppress-format: true
```
You don't HAVE to group them. It's just a matter of coding style

**Review by isra-fel — CHANGES_REQUESTED** [review]

File: src/StorageMover/examples/Update-AzStorageMoverProject.md (hunk 9 above)
Comment (yifanz0) [reply src/StorageMover/examples/Update-AzStorageMoverProject.md]: Done.

File: src/StorageMover/examples/New-AzStorageMoverJobDefinition.md (hunk 10 above)
Comment (yifanz0) [reply src/StorageMover/examples/New-AzStorageMoverJobDefinition.md]: Done

File: src/StorageMover/README.md (hunk 11 above)
Comment (yifanz0) [reply src/StorageMover/README.md]: Thanks for the suggestion. Yes this looks much cleaner than listing them out one by one. Fixed.

**Review by isra-fel — APPROVED** [review]

**Pull request merged**